Add blog about Node Log Query alpha (#39829)
kubernetes/website: content/en/blog/_posts/2023-03-06-node-log-viewer-alpha.md

---
layout: blog

> title: "Node Log Viewer"

Reviewer: ```suggestion
title: "Kubernetes 1.27: Access Node Logs Using The Kubernetes API"
```
“viewer” implies there's a client tool available.

Author: I renamed this to `Kubernetes 1.27: Query Node Logs Using The Kubelet API` as this is a not a first class Kubernetes API that goes through the API server but is just wired directly through the `proxy/log` endpoint in the kubelet.

logs of the service to debug the issue. The _Node log query_ feature helps with
this scenario by allowing the cluster administrator to view the logs using
_kubectl_. This is especially useful with Windows nodes where you run into the
issue of the node going to the ready state but containers not coming up due to
CNI misconfigurations and other issues that are not easily identifiable by
looking at the Pod status.

## How does it work?

The kubelet already has a _/var/log/_ viewer that is accessible via the node
proxy endpoint. The feature supplements this endpoint with a shim that shells
out to _journalctl_, on Linux nodes, and _Get-WinEvent_ cmdlet on Windows
nodes. It then uses the existing filters provided by the commands to allow
filtering the logs. The kubelet also uses heuristics to retrieve the logs.
If the user is not aware if a given system services logs to a file or to the
native system logger, the heuristics first checks the native operating system
logger and if that is not available it attempts to retrieve the first logs
from `/var/log/<servicename>` or `/var/log/<servicename>.log` or
`/var/log/<servicename>/<servicename>.log`.

> On Linux we assume that service logs are available via journald. On Windows we
> assume that service logs are available in the application log provider. Note
> that this feature requires you to have access to node objects.

Reviewer: ```suggestion
On Linux we assume that service logs are available via journald, and that
`journalctl` is installed. On Windows we assume that service logs are available in the
application log provider. Also note that fetching node logs is only available if you are authorized
to do so (in RBAC, that's **get** and **create** access to `nodes/proxy`). The privileges that you
need to fetch node logs also allow elevation-of-privilege attacks, so be careful about how you 
manage them.
```

Reviewer 2: I'd agree that cluster operators will need to be very careful if the permission required is create and get on node/proxy. Those rights allow users to directly contact the kubelet API (bypassing auditing and admission control) and execute commands in any pod on a node (amongst other things)

Author: I want to call out that the ability read logs using the `/proxy/logs` endpoint is already available and enabled by default. This feature is extending it be able to read journald logs and defaults to false.

## How do I use it?

To use the feature, ensure that the `NodeLogQuery`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) is
enabled for that node, and that the kubelet configuration options
`enableSystemLogHandler` and `enableSystemLogQuery` are both set to true. You can
then query the logs from all your nodes or just a subset. Here is an example to
retrieve the kubelet service logs from a node:
```shell
# Fetch kubelet logs from a node named node-1.example
kubectl get --raw "/api/v1/nodes/node-1.example/proxy/logs/?query=kubelet"
```

You can further filter the query to narrow down the results:
```shell
# Fetch kubelet logs from a node named node-1.example that have the word "error"
kubectl get --raw "/api/v1/nodes/node-1.example/proxy/logs/?query=kubelet&pattern=error"
```

You can also fetch files from `/var/log/` on a Linux node:
```shell
kubectl get --raw "/api/v1/nodes/<insert-node-name-here>/proxy/logs/?query=/<insert-log-file-name-here>"
```

You can read the

> [documentation](content/en/docs/concepts/cluster-administration/system-logs.md#log-query)

Reviewer: ```suggestion
[documentation](/en/docs/concepts/cluster-administration/system-logs/#log-query)
```
